File: internal/engine/memtable.go
package engine

import "sync"

const skipListNodeOverhead = 130

type MemTable struct {
	mu        sync.RWMutex
	data      *SkipList
	sizeBytes int
}

func NewMemTable() *MemTable {
	return &MemTable{
		data:      NewSkipList(),
		sizeBytes: 0,
	}
}

func (m *MemTable) Put(userKey []byte, value []byte, seqNum uint64) {
	m.mu.Lock()
	defer m.mu.Unlock()

	internalKey := EncodeInternalKey(userKey, seqNum, TypePut)

	m.data.Insert(internalKey, value)

	m.sizeBytes += len(internalKey) + len(value) + skipListNodeOverhead
}

func (m *MemTable) Get(userKey []byte, targetSeqNum uint64) ([]byte, bool, bool) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	searchKey := EncodeInternalKey(userKey, targetSeqNum, TypePut)

	val, keyType, found := m.data.Search(searchKey)

	if !found {
		return nil, false, false
	}

	if keyType == TypeDelete {
		return nil, true, true
	}

	return val, false, true
}

func (m *MemTable) Delete(userKey []byte, seqNum uint64) {
	m.mu.Lock()
	defer m.mu.Unlock()

	internalKey := EncodeInternalKey(userKey, seqNum, TypeDelete)

	m.data.Insert(internalKey, nil)

	m.sizeBytes += len(internalKey) + skipListNodeOverhead
}

func (m *MemTable) Iterate(cb func(internalKey []byte, value []byte)) {
	m.mu.RLock()
	defer m.mu.RUnlock()
	m.data.Iterate(cb)
}

func (m *MemTable) ApproximateSize() int {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return m.sizeBytes
}

File: internal/engine2/Memtable.cpp
#include "MemTable.hpp"

MemTable::MemTable() : data(new SkipList()), sizeBytes(0) {}

MemTable::~MemTable() {
    delete data;
}

void MemTable::Put(const std::string& userKey, const std::string& value, uint64_t seqNum) {
    std::unique_lock<std::shared_mutex> lock(mu);
    
    std::string internalKey = EncodeInternalKey(userKey, seqNum, TypePut);
    data->Insert(internalKey, value);
    
    sizeBytes += internalKey.size() + value.size() + skipListNodeOverhead;
}

std::tuple<std::string, bool, bool> MemTable::Get(const std::string& userKey, uint64_t targetSeqNum) {
    std::shared_lock<std::shared_mutex> lock(mu);
    
    std::string searchKey = EncodeInternalKey(userKey, targetSeqNum, TypePut);
    auto [val, keyType, found] = data->Search(searchKey);

    if (!found) {
        return {"", false, false};
    }

    if (keyType == TypeDelete) {
        return {"", true, true};
    }

    return {val, false, true};
}

void MemTable::Delete(const std::string& userKey, uint64_t seqNum) {
    std::unique_lock<std::shared_mutex> lock(mu);
    
    std::string internalKey = EncodeInternalKey(userKey, seqNum, TypeDelete);
    data->Insert(internalKey, "");
    
    sizeBytes += internalKey.size() + skipListNodeOverhead;
}

void MemTable::Iterate(std::function<void(const std::string&, const std::string&)> cb) {
    std::shared_lock<std::shared_mutex> lock(mu);
    data->Iterate(cb);
}

size_t MemTable::ApproximateSize() {
    std::shared_lock<std::shared_mutex> lock(mu);
    return sizeBytes;
}